File: frontend/src/lib/api/diagnostic.ts
import { auth } from '$lib/stores/auth.svelte';

export interface DiagnosticSession {
  id: number;
  user_id: number;
  materia_id: number;
  numero_intento: number;
  estado: string;
  estrategia: any;
  preguntas_totales: number;
  preguntas_correctas: number;
  started_at: string;
  completed_at?: string;
}

export interface DiagnosticQuestion {
  id: number;
  oa_bloom_objective_id: number;
  tipo: string;
  question_data: any;
  question_number: number;
  total_questions: number;
  current_bloom_level: number;
}

export interface DiagnosticAnswer {
  is_correct: boolean;
  score: number;
  answer_id: number;
  new_bloom_level: number;
}

export interface DiagnosticResult {
  id: number;
  session_id: number;
  oa_id: number;
  nivel_bloom_dominado: number;
  nivel_bloom_nombre: string;
  preguntas_respondidas: number;
  preguntas_correctas: number;
  porcentaje_aciertos: number;
  recomendacion: string;
  oa?: {
    id: number;
    codigo: string;
    titulo: string;
    descripcion: string;
  };
}

export interface DiagnosticCompleteResponse {
  message: string;
  session: DiagnosticSession;
  average_bloom_level: number;
}

const API_BASE_URL = '/api';

function getAuthHeaders() {
  const token = auth.token;
  return {
    'Content-Type': 'application/json',
    'Authorization': token ? `Bearer ${token}` : ''
  };
}

export async function startDiagnosticSession(materiaId: number) {
  try {
    const response = await fetch(`${API_BASE_URL}/diagnostic-sessions`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({ materia_id: materiaId })
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const session: DiagnosticSession = await response.json();
    return { success: true, session };
  } catch (error) {
    console.error('Error starting diagnostic session:', error);
    return { success: false, error: 'Network error' };
  }
}

export async function getNextQuestion(sessionId: number) {
  try {
    const response = await fetch(`${API_BASE_URL}/diagnostic-sessions/${sessionId}/next-question`, {
      method: 'GET',
      headers: getAuthHeaders()
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const question: DiagnosticQuestion = await response.json();
    return { success: true, question };
  } catch (error) {
    console.error('Error getting next question:', error);
    return { success: false, error: 'Network error' };
  }
}

export async function submitAnswer(sessionId: number, questionId: number, userAnswer: any, tiempoSegundos?: number) {
  try {
    const response = await fetch(`${API_BASE_URL}/diagnostic-sessions/${sessionId}/answer`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({
        question_id: questionId,
        user_answer: userAnswer,
        tiempo_segundos: tiempoSegundos
      })
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const result: DiagnosticAnswer = await response.json();
    return { success: true, result };
  } catch (error) {
    console.error('Error submitting answer:', error);
    return { success: false, error: 'Network error' };
  }
}

export async function completeSession(sessionId: number) {
  try {
    const response = await fetch(`${API_BASE_URL}/diagnostic-sessions/${sessionId}/complete`, {
      method: 'POST',
      headers: getAuthHeaders()
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const result: DiagnosticCompleteResponse = await response.json();
    return { success: true, result };
  } catch (error) {
    console.error('Error completing session:', error);
    return { success: false, error: 'Network error' };
  }
}

export async function getResults(sessionId: number) {
  try {
    const response = await fetch(`${API_BASE_URL}/diagnostic-sessions/${sessionId}/results`, {
      method: 'GET',
      headers: getAuthHeaders()
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const results: DiagnosticResult[] = await response.json();
    return { success: true, results };
  } catch (error) {
    console.error('Error getting results:', error);
    return { success: false, error: 'Network error' };
  }
}

export async function getSessionProgress(sessionId: number) {
  try {
    const response = await fetch(`${API_BASE_URL}/diagnostic-sessions/${sessionId}`, {
      method: 'GET',
      headers: getAuthHeaders()
    });

    if (!response.ok) {
      const error = await response.text();
      return { success: false, error };
    }

    const session: DiagnosticSession = await response.json();
    return { success: true, session };
  } catch (error) {
    console.error('Error getting session progress:', error);
    return { success: false, error: 'Network error' };
  }
}

File: backend/internal/handlers/diagnostic.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/platanus-hack-25/lumera_app/internal/db"
	authmiddleware "github.com/platanus-hack-25/lumera_app/internal/middleware"
	"github.com/platanus-hack-25/lumera_app/internal/models"
	"github.com/platanus-hack-25/lumera_app/internal/services"
	"gorm.io/datatypes"
)

// StartDiagnosticRequest represents the request to start a diagnostic session
type StartDiagnosticRequest struct {
	MateriaID uint `json:"materia_id"`
}

// StartDiagnostic godoc
// @Summary Start a diagnostic session
// @Description Create a new diagnostic session for a student
// @Tags Diagnostic
// @Accept json
// @Produce json
// @Param diagnostic body StartDiagnosticRequest true "Diagnostic session data"
// @Success 201 {object} models.DiagnosticSession
// @Failure 400 {object} map[string]interface{}
// @Security BearerAuth
// @Router /api/diagnostic-sessions [post]
func StartDiagnostic(w http.ResponseWriter, r *http.Request) {
	userID, ok := authmiddleware.GetUserIDFromContext(r.Context())
	if !ok {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	var req StartDiagnosticRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Count previous attempts
	var count int64
	db.DB.Model(&models.DiagnosticSession{}).
		Where("user_id = ? AND materia_id = ?", userID, req.MateriaID).
		Count(&count)

	// Initialize adaptive strategy
	defaultStrategy := map[string]interface{}{
		"nivel_bloom_actual":      2,
		"oas_evaluados":           []uint{},
		"aciertos_consecutivos":   0,
		"fallos_consecutivos":     0,
		"patron_respuestas":       []string{},
	}
	strategyJSON, _ := json.Marshal(defaultStrategy)

	session := models.DiagnosticSession{
		UserID:        userID,
		MateriaID:     req.MateriaID,
		NumeroIntento: int(count) + 1,
		Estado:        "en_progreso",
		Estrategia:    datatypes.JSON(strategyJSON),
		StartedAt:     time.Now(),
	}

	if err := db.DB.Create(&session).Error; err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(session)
}

// GetSessionProgress godoc
// @Summary Get diagnostic session progress
// @Description Retrieve diagnostic session details and progress
// @Tags Diagnostic
// @Produce json
// @Param id path int true "Session ID"
// @Success 200 {object} models.DiagnosticSession
// @Failure 404 {object} map[string]interface{}
// @Security BearerAuth
// @Router /api/diagnostic-sessions/{id} [get]
func GetSessionProgress(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	var session models.DiagnosticSession

	if err := db.DB.Preload("Materia").Preload("Answers").First(&session, id).Error; err != nil {
		http.Error(w, "Session not found", http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(session)
}

// SubmitAnswerRequest represents the request to submit an answer
type SubmitAnswerRequest struct {
	QuestionID         uint           `json:"question_id"`
	UserAnswer         datatypes.JSON `json:"user_answer"`
	TiempoSegundos     *int           `json:"tiempo_segundos"`
}

// SubmitAnswer godoc
// @Summary Submit an answer during diagnostic
// @Description Submit an answer and get validation result
// @Tags Diagnostic
// @Accept json
// @Produce json
// @Param id path int true "Session ID"
// @Param answer body SubmitAnswerRequest true "Answer data"
// @Success 200 {object} map[string]interface{}
// @Failure 400 {object} map[string]interface{}
// @Security BearerAuth
// @Router /api/diagnostic-sessions/{id}/answer [post]
func SubmitAnswer(w http.ResponseWriter, r *http.Request) {
	sessionID := chi.URLParam(r, "id")
	var session models.DiagnosticSession

	if err := db.DB.First(&session, sessionID).Error; err != nil {
		http.Error(w, "Session not found", http.StatusNotFound)
		return
	}

	var req SubmitAnswerRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Get question
	var question models.Question
	if err := db.DB.Preload("OABloomObjective").First(&question, req.QuestionID).Error; err != nil {
		http.Error(w, "Question not found", http.StatusNotFound)
		return
	}

	// Validate answer
	isCorrect, score, err := question.ValidateAnswer(req.UserAnswer)
	if err != nil && err.Error() != "requires manual or AI validation" {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Save answer
	answer := models.DiagnosticAnswer{
		SessionID:          session.ID,
		QuestionID:         req.QuestionID,
		OABloomObjectiveID: question.OABloomObjectiveID,
		BloomLevelID:       question.OABloomObjective.BloomLevelID,
		UserAnswer:         req.UserAnswer,
		IsCorrect:          &isCorrect,
		Score:              &score,
		TiempoSegundos:     req.TiempoSegundos,
	}

	if err := db.DB.Create(&answer).Error; err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Update session stats
	session.PreguntasTotales++
	if isCorrect {
		session.PreguntasCorrectas++
	}
	db.DB.Save(&session)

	response := map[string]interface{}{
		"is_correct": isCorrect,
		"score":      score,
		"answer_id":  answer.ID,
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

// CompleteDiagnostic godoc
// @Summary Complete a diagnostic session
// @Description Mark session as completed and generate results
// @Tags Diagnostic
// @Produce json
// @Param id path int true "Session ID"
// @Success 200 {object} map[string]interface{}
// @Failure 400 {object} map[string]interface{}
// @Security BearerAuth
// @Router /api/diagnostic-sessions/{id}/complete [post]
func CompleteDiagnostic(w http.ResponseWriter, r *http.Request) {
	sessionID := chi.URLParam(r, "id")
	var session models.DiagnosticSession

	if err := db.DB.First(&session, sessionID).Error; err != nil {
		http.Error(w, "Session not found", http.StatusNotFound)
		return
	}

	now := time.Now()
	session.Estado = "completado"
	session.CompletedAt = &now

	if err := db.DB.Save(&session).Error; err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// TODO: Generate diagnostic_results based on answers
	// This would analyze answers and create DiagnosticResult records
	// which will trigger the auto-update of student_oa_progress

	// Gamification events (async)
	go func() {
		// Award XP for completing diagnostic
		gamificationService.AddXP(session.UserID, 100, "diagnostic_completed")

		// Bonus coins for good performance
		if session.PreguntasTotales > 0 {
			scorePercent := (session.PreguntasCorrectas * 100) / session.PreguntasTotales
			if scorePercent >= 80 {
				gamificationService.AddCoins(session.UserID, 50, "diagnostic_excellent")
			} else if scorePercent >= 60 {
				gamificationService.AddCoins(session.UserID, 25, "diagnostic_good")
			}

			// Check for diagnostic achievement unlocks
			unlockService.CheckAndUnlock(session.UserID, services.UnlockEvent{
				Type: "diagnostic_achievement",
				Key:  fmt.Sprintf("diagnostic_materia_%d_score_%d", session.MateriaID, scorePercent),
				Data: map[string]interface{}{
					"materia_id": session.MateriaID,
					"score":      scorePercent,
				},
			})
		}
	}()

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"message": "Diagnostic completed successfully",
		"session": session,
	})
}

// GetDiagnosticResults godoc
// @Summary Get diagnostic results
// @Description Retrieve consolidated results for a completed diagnostic session
// @Tags Diagnostic
// @Produce json
// @Param id path int true "Session ID"
// @Success 200 {array} models.DiagnosticResult
// @Failure 404 {object} map[string]interface{}
// @Security BearerAuth
// @Router /api/diagnostic-sessions/{id}/results [get]
func GetDiagnosticResults(w http.ResponseWriter, r *http.Request) {
	sessionID := chi.URLParam(r, "id")
	var results []models.DiagnosticResult

	if err := db.DB.Where("session_id = ?", sessionID).
		Preload("OA").
		Find(&results).Error; err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(results)
}
